x/qgb/abci.go

```diff
@@ -114,3 +122,67 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
 		}
 	}
 }
+
+// maybePruneAttestations runs basic checks on saved attestations to see if we need to prune or not.
+// Then, it starts pruning the expired attestations until all attestations in state are still valid.
+func maybePruneAttestations(ctx sdk.Context, k keeper.Keeper) {
+	// If the attestations nonce hasn't been initialized yet, no pruning is
+	// required
+	if !k.CheckLatestAttestationNonce(ctx) {
+		return
+	}
+	earliestAttestation, found, err := k.GetAttestationByNonce(ctx, k.GetEarliestAvailableAttestationNonce(ctx))
+	if err != nil {
+		ctx.Logger().Error("error getting earliest attestation for pruning", "err", err.Error())
+		return
+	}
+	if !found {
+		ctx.Logger().Error("couldn't find earliest attestation for pruning")
+		return
+	}
+	if earliestAttestation == nil {
+		ctx.Logger().Error("nil earliest attestation")
+		return
+	}
+	currentBlockTime := ctx.BlockTime()
+	// if the current time is before the earliest attestation creation time + expiry time
+	// then, all the subsequent attestations are also still valid and no need to prune them.
+	if currentBlockTime.Before(earliestAttestation.BlockTime().Add(AttestationExpiryTime)) {
+		return
+	}
+
+	ctx.Logger().Debug("pruning attestations from QGB store")
+	latestAttestationNonce := k.GetLatestAttestationNonce(ctx)
+	var newEarliestAvailableNonce uint64
+	for newEarliestAvailableNonce = earliestAttestation.GetNonce(); newEarliestAvailableNonce < latestAttestationNonce; newEarliestAvailableNonce++ {
+		newEarliestAttestation, found, err := k.GetAttestationByNonce(ctx, newEarliestAvailableNonce)
+		if err != nil {
+			ctx.Logger().Error("error getting attestation for pruning", "nonce", newEarliestAvailableNonce, "err", err.Error())
+			return
+		}
+		if !found {
+			ctx.Logger().Error("couldn't find attestation for pruning", "nonce", newEarliestAvailableNonce)
+			return
+		}
+		if newEarliestAttestation == nil {
+			ctx.Logger().Error("nil attestation for pruning", "nonce", newEarliestAvailableNonce)
+			return
+		}
+		if currentBlockTime.Before(newEarliestAttestation.BlockTime().Add(AttestationExpiryTime)) {
+			// the earliest attestation is still valid. this means all the subsequent ones are also
```

Comment:
same comment as above, I think adding then comparing before is if just checking if a < b with extra steps

Reply:
[7986e04](https://github.com/celestiaorg/celestia-app/pull/1799/commits/7986e04c63eb7a419646978fe36ebe714c65291f)